Merge branch 'master' into feature1

mirik123 opened the pull request:
This is a new completely rewritten Visual Studio plugin utilizing the QuantConnect REST API to launch backtests and display results inside Visual Studio 2012-2015.

This plugin consists of 3 parts:
1. The main VSIX project
2. The Algorithm project template
3. The Nuget package that contains Lean binaries necessary for Algorithm project <br/>(can be removed when you implement your own nuget package)

### The main changes

VSIX project
1. Using WebHttpRequest instead of the RestSharp (RestSharp is not strong signed).
2. Using async/await instead of the ConcurrentQueue.
3. Using WindowToolPane instead of the WinForms window.
4. Al logics responsible for chart draw has been moved to a separate module.
5. Removed unnecessary dialogs.
6. Corrected and optimized workflow.

Algorithm project template
1. Using Visual Studio project template instead of the downloading it from GitHub.
2. Using Lean binaries from Nuget package instead of the downloading them from GitHub.

### Features left TODO

1. Spread components to separate WindowToolPanes:
   1. Charts
   2.  Statistics
   3. Administration (Projects/Backtests)
   4. Trades
2. Create new dialogs to allow Uploading/Downloading only a subset of project files.
3. Use commercial high-grade framework for drawing stock charts (like SciChart) instead of the ZedGraph.

jaredbroad commented:
@mirik123 nice work! I like where you're going, if you're interested we can whip this into shape and get it ready for a merge.

 - Lets move as many of the DLL's as possible into NuGet calls, there's a lot of DLL's committed here which we can avoid committing to the repo.

 - The QCAlgorithm base class is now in the LEAN repo, I saw you submitted a Nuget package for LEAN -- thats a great idea, we can move many of these shared DLL's into NuGet to allow easy usage.

 - The LEAN repo now includes a HTML charting API. We can actually do full local charting using our powerful JS charting library. This is fully featured and will be kept up to date unlike anything SciChart/ZedGraph can do. So ideally we should make the charting link up with that. Backtests running in our cloud will still stream to a local plugin like on the website. Check out the "Interface" project in LEAN to see the example, using the "backtest-desktop" environment

mirik123 commented:
### Release notes:

1) Added JavaScript web-side chating solution like used in Lean project [QuantConnect.Views](https://github.com/QuantConnect/Lean/tree/master/UserInterface/WinForms)
2) Removed QCAlgorithm project template. There are updated project templates in Lean project [QuantConnect.Algorithm.CSharp](https://github.com/QuantConnect/Lean/tree/master/Algorithm.CSharp)
3) Removed NuGet dependency of Lean binaries. It is better to take them by recompiling Lean project [QuantConnect](https://github.com/QuantConnect/Lean/tree/master/Common)

jaredbroad commented:
Pretty cool! :) Getting close to mergability; please remove comment headers and replace with the LEAN open source APACHE license headers. Please also remove all "mnFunction" style naming conventions and replace with MS convention. We have a contributing guide [here](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).

jaredbroad commented:
When loading the JS chart you need to trigger a JS function which tells the page to load - you can see more about this here: https://github.com/QuantConnect/Lean/blob/master/UserInterface/WinForms/LeanWinForm.cs#L189

This allows users to inject the results they want the chart to display into a variable on the page, and then call the "ready" function to display those results (for static past backtests)

jaredbroad commented:
Theres some pretty exciting potential there. Ideally we wouldn't have all the QC classes duplicated in this repo (or maintenance will become difficult). With a substantial tidy up we could consider moving this into the main LEAN repo.

Please look at the API v2 upgrades in the LEAN repo - the v1 will be deprecated soon.
https://github.com/QuantConnect/Lean/blob/master/Api/Api.cs

I'll leave this open but I'm not sure about merging it in. This plugin is deprecated, we'd need a solid redesign to consider opening it again.

mirik123 commented:
The ability to develop the code in Visual Studio IDE is the **kill-feature**.
There are a lot of other platforms for algorithmic trading but no one of them integrates with Visual Studio (except Eclipse oriented frameworks).  Moreover, most of them invent a bicycle and use their own "unique" scripts.
Combine QCStudioPlugin with Lean is a good idea. Till now the main problem with Lean that I've experienced was that the Lean is not strong-signed and it also uses other 3rd party libraries that are not strong signed too. Without strong-singing Lean cannot be integrated with Visual Studio environment.

mchandschuh commented:
Just a cursory [search](http://stackoverflow.com/a/16452046/1582922) shows deployable .vsix extensions don't require strong names, but the VS wizards may place that requirement. So maybe there is a way to do it without strong naming all of our assemblies? (and third party for which we have no control).
